cpp/src/strings/convert/convert_durations.cu

```diff
@@ -0,0 +1,650 @@
+ */
+struct format_compiler {
+  std::string format;
+  rmm::device_vector<format_item> d_items;
+
+  std::map<char, int8_t> specifier_lengths = {{'d', -1},
+                                              {'+', -1},  // only for negative days
+                                              {'H', 2},
+                                              {'M', 2},
+                                              {'S', 2},
+                                              {'u', -1},   // 0 or 6+1(dot)
+                                              {'f', -1}};  // 0 or <=9+1(dot) without trialing zeros
+
+  format_compiler(const char* format) : format(format) {}
+
+  format_item const* compile_to_device()
+  {
+    std::vector<format_item> items;
+    const char* str         = format.c_str();
+    auto length             = format.length();
+    const bool is_isoformat = str[0] == 'P';
+    if (is_isoformat) {
+      specifier_lengths['H'] = -1;
+      specifier_lengths['M'] = -1;
+      specifier_lengths['S'] = -1;
+    }
+    while (length > 0) {
+      char ch = *str++;
+      length--;
+      if (ch != '%') {
+        items.push_back(format_item::new_delimiter(ch));
+        continue;
+      }
+      CUDF_EXPECTS(length > 0, "Unfinished specifier in duration format");
+
+      ch = *str++;
+      length--;
+      if (ch == '%')  // escaped % char
+      {
+        items.push_back(format_item::new_delimiter(ch));
+        continue;
+      }
+      if (ch >= '0' && ch <= '9') {
+        CUDF_EXPECTS(*str == 'f', "precision not supported for specifier: " + std::string(1, *str));
+        specifier_lengths[*str] = static_cast<int8_t>(ch - '0') + 1;  // +1 is for dot
+        ch                      = *str++;
+        length--;
+      }
+      CUDF_EXPECTS(specifier_lengths.find(ch) != specifier_lengths.end(),
+                   "invalid format specifier: " + std::string(1, ch));
+
+      int8_t spec_length = specifier_lengths[ch];
+      items.push_back(format_item::new_specifier(ch, spec_length));
+    }
+    // create program in device memory
+    d_items.resize(items.size());
+    CUDA_TRY(cudaMemcpyAsync(
+      d_items.data().get(), items.data(), items.size() * sizeof(items[0]), cudaMemcpyHostToDevice));
+    return d_items.data().get();
+  }
+
+  // these calls are only valid after compile_to_device is called
+  size_type items_count() const { return static_cast<size_type>(d_items.size()); }
+};
+
+__device__ void dissect_duration(int64_t duration, int32_t* timeparts, type_id units)
+{
+  if (units == type_id::DURATION_DAYS) {
+    timeparts[DU_DAY] = static_cast<int32_t>(duration);
+    return;
+  }
+
+  // Specialized modulo expression that handles negative values.
+  // Examples:
+  //     modulo(1,60)    1
+  //     modulo(-1,60)  59
+  auto modulo_time = [](int64_t time, int64_t base) {
+    return static_cast<int32_t>(((time % base) + base) % base);
+  };
+
+  duration_s seconds{0};
+  if (units == type_id::DURATION_SECONDS) {
```

Comment:
If `dissect_duration` was instead a template function that operated on a `duration` type instead of `int64_t`, then it could be specialized instead of using the `units` runtime information. Then I believe the access pattern to `timeparts` would be all known at compile time, and therefore be done in registers instead of local memory.

Reply:
done. `units` is removed.